### DISCUSSION TITLE
equality operator overload on objects

### DISCUSSION THREAD

**Discussion by bytezeroseven** [quickjs-ng/quickjs#365] [Q&A]
How will I go about overloading equality operator on two objects? Both operands will be objects. I know how to do it when one operand is a primitive. Can someone provide me intructions as to what I should do in the C files?

**Comment by saghul** [comment]
I don't think that is currently possible.

**Comment by bnoordhuis** [comment]
More specifically, objects have identity so overloading equality doesn't make sense. At that point it's no longer javascript.

**Comment by bytezeroseven** [comment]


**Reply by saghul** [reply]
Maybe use a WeakMap?

**Reply by bytezeroseven** [reply]
can i instead somehow do it with storing objects on heap? basically a C function should return an object from an ID. 
get_object(id) can be a C function exposed to JS. It creates or returns cached objects based on an ID. 
I have tried by storing a pointer on heap and then returning `*ptr` but it throws this error. 

`Aborted(Assertion failed: p->ref_count == 0, at: ./quickjs/quickjs.c,5446,free_zero_refcount)
RuntimeError: Aborted(Assertion failed: p->ref_count == 0, at: ./quickjs/quickjs.c,5446,free_zero_refcount). Build with -sASSERTIONS for more info.`

**Reply by bnoordhuis** [reply]
Storing objects as JSValues in a C side data structure should work, you just have to JS_FreeValue them all before you destroy the JSContext. Those assertions suggest you don't do that.

**Reply by bytezeroseven** [reply]
I think keeping it on heap is preventing natural garbage collection from quickjs side. any way to fix that?

**Reply by bnoordhuis** [reply]
You're almost certainly doing something you shouldn't be doing if you have to adjust reference counts manually but it's not entirely clear to me what. What does "keeping it on the heap" mean in detail? Can you post your code?

**Reply by saghul** [reply]
Some APIs "consume" a reference, so it would be decreased.
